Units rework II

3x380V opened the pull request:
This just a  #18844 rebased on the top of #20540. Units were _not_ moved to `Units` namespace as it introduced a lot of changes everywhere. Still might be done later. The rest of changes are based on review comments and testing.

kadet1090 commented on src/Base/Unit.cpp:
```diff
@@ -27,696 +26,243 @@
-    memcpy(&sig, &Val, sizeof(Val));
-    auto isInt = [](double value) {
-        return std::fabs(std::round(value) - value) < std::numeric_limits<double>::epsilon();
+    auto mult = [&](auto leftVal, auto rightVal) {
```
Just a thought - maybe it should be called `leftExponent` and `rightExponent`? I am afraid that someone will try to "fix" that function by changing + to * as it is named mult.

3x380V replied:
...but you would not allow such change go in, right? Also unit test would fail. I have no strong feelings here, but prefer short and simple name, so I'm fine with current naming. Of course if that is show stopper I will change it.

kadet1090 replied:
Not show stopper, but it may create confusion for people reading it which we should avoid. Of course I'm not gonna hold PR over such minor detail - but it would be nice to change I think.

3x380V commented:
I was thinking about template encapsulating whole
```
    UnitVals res {};
    std::transform(_vals.begin(), _vals.end(), right._vals.begin(), res.begin(), div);
    return Unit {res};
```
snippet, but... Could we focus on the first 7 commits now?

hyarion commented on src/App/FeatureTest.cpp:
```diff
@@ -145,7 +145,7 @@ FeatureTest::FeatureTest()
     QuantityLength.setUnit(Base::Unit::Length);
     ADD_PROPERTY(QuantityOther, (5.0));
-    QuantityOther.setUnit(Base::Unit(-3, 1));
+    QuantityOther.setUnit(Base::Unit::Density);
```
To me the original code seemed to test that it is possible to set a custom unit that does not exist yet in FreeCAD. I think the property would have been named `QuantityDensity` otherwise. What do you think here?

3x380V replied:
That was added by c5316d31a92f ("add quantity property to FeatureTest with kg/m^3"), previous quantities were added by 44c07da571d1 ("integrate Quantities into PropertyView and activate preferences again").
Density unit was added by 7d8fb99cea ("Units: Add Density to internal schema") and it was done a year before that test was added. Honestly you might be right and it is a pity @wwmayer left for the time being.

Anyway, my motivation for this change is different Unit constructor after refactoring, so it avoids changing this test. Constructing Unit itself is covered by tests/src/Base/Unit.cpp.

hyarion replied:
c5316d31a92fcdfe79cf97b6d6de43233a21a52d was added right after 7544444d638661a63485c8149b0e96b1334e47cd ("issue [#0002625](https://github.com/benj5378/FreeCAD/issues/0002625): Wrong dimension (str) from Unit") which is fixing the issue from here: https://tracker.freecad.org/view.php?id=2625

I think you might be right and that this test was just created like this to specifically test using the base quantities for fixing that bug.
Given this context I don't think it matters anymore. I'm marking this as resolved.

hyarion commented on src/Mod/Sketcher/Gui/EditDatumDialog.cpp:
```diff
@@ -201,7 +201,8 @@ int EditDatumDialog::exec(bool atCursor)
 {
     Base::Quantity newQuant = ui_ins_datum->labelEdit->value();
-    if (newQuant.isQuantity() || (Constr->Type == Sketcher::SnellsLaw && newQuant.isDimensionless())
+    if (!newQuant.isDimensionless()
+        || (Constr->Type == Sketcher::SnellsLaw && newQuant.isDimensionless())
```
`isQuantity` is changed to `isDimensionless` here which makes the expression a bit strange compared to the old one.

As we're already checking if `!newQuant.isDimensionless()` this can be simplified to:
```cpp
    if (!newQuant.isDimensionless()
        || Constr->Type == Sketcher::SnellsLaw
        || Constr->Type == Sketcher::Weight) {
```

I don't know if that is correct (it does look sound), but maybe you could double check it as you have better domain knowledge?

hyarion reviewed (commented):
6804c5b24a9ebb775c9f5630155c0d5baff32314: ✅ 
2635a9ee971d81072b7a000c63859b1f1f699cd2: ✅
5726d019222d760e2b25a7ee662f5f31b7e9e4c9: ✅
f5ebaa32e11e957150a0782e668de077ecb39747: added comments in commit
55fe2e8e6f498c7fafd962ab72f7d2195d833ab2: added comments in commit
48eb633d46263be28d95e72a956b853cc5351b51: ✅ 
6cb969cba1fe4d4aefa08832847f47719686e7db: ✅

benj5378 commented on src/Base/Unit.h:
```diff
@@ -1,179 +1,260 @@
 
-inline Unit& Unit::operator*=(const Unit& that)
+constexpr auto numUnitSpecs {58};
+constexpr std::array<std::pair<std::string_view, UnitVals>, numUnitSpecs> unitSpecs {{
```
Consider named struct instesad of std::pair

3x380V replied:
That for sure I want to avoid to save us yet another round of comments considering this and that name in that named struct.

kadet1090 replied:
Wouldn't it just be:

```cpp
struct UnitSpec { 
    std::string_view name;
    UnitVals vals;
}
```
   
You are already using that exact names (although, with _ prefix) inside the `Unit` class. I don't think that there will be much debate about the names.

Another thing is - does it need to be `std::array`? With `std::array` you need to specify the number of entries manually, which is redundant. `std::initializer_list` should be a good alternative for that use case as it does not introduce much overhead (if any) and it does not need such redundancy.

3x380V replied:
Is `std::initializer_list` `constexpr` in C++20? Note, code was written at the time FreeCAD code base was at C++17.

kadet1090 replied:
https://en.cppreference.com/w/cpp/utility/initializer_list/initializer_list based on that it seems that initializer_list is constexpr since c++14, I also did not have a problem with using it recently.

3x380V replied:
Okay, I cannot test it right now, but give it a try once back in the office.
However, so far we are doing quite well according the Law of Triviality. There are more fundamental problems with units being there since the beginning. For example, current unit system is unable to guess name of computed unit, it just merely returns the last one defined. Comparing to this (having to define units in the right order, so tests pass) is specifying the number of entries just a detail... Question is whenever we want to fix that.

benj5378 commented on src/Base/Unit.h:
```diff
@@ -1,179 +1,260 @@
-    *this = *this * that;
-    return *this;
+    auto match = [&](const std::pair<std::string_view, UnitVals>& pair) {
+        return pair.first == str;
+    };
+
+    const auto found = std::find_if(unitSpecs.begin(), unitSpecs.end(), match);
+    return found != unitSpecs.end();
 }
 
-inline Unit& Unit::operator/=(const Unit& that)
+inline std::string getString(const UnitVals vals)
 {
-    *this = *this / that;
-    return *this;
+    auto match = [&](const std::pair<std::string_view, UnitVals>& pair) {
+        return pair.second == vals;
+    };
+
+    const auto found = std::find_if(unitSpecs.begin(), unitSpecs.end(), match);
+    return std::string(found == unitSpecs.end() ? "" : found->first);
```
I would suggest moving this to .cpp, to not have `#include <algorithm>" in the whole codebase

hyarion replied:
Right, this specific function uses `algorithm` so it makes sense to drop `inline` and move the implementation to the cpp file.

3x380V commented:
@hyarion, I have dropped these two for now:
> [2635a9e](https://github.com/FreeCAD/FreeCAD/commit/2635a9ee971d81072b7a000c63859b1f1f699cd2): ✅ 
> [5726d01](https://github.com/FreeCAD/FreeCAD/commit/5726d019222d760e2b25a7ee662f5f31b7e9e4c9): ✅ 

The intent was to avoid spaghetti from UnitApi to Quantity and backwards, but without moving float formatting stuff around `QuantityFormat` elsewhere that won't work. I do not want to further complicate this PR now.

hyarion commented:
@3x380V It looks like we're getting ready soon for merge. May I ask what's left before it's pulled out from Draft status?

3x380V commented:
I'd like to test `std::initializer_list` idea and of course give it some more testing.

3x380V commented:
@bofdahof, thanks, that works.

3x380V commented on src/Base/QuantityPyImp.cpp:
```diff
@@ -40,6 +40,8 @@
 
 using Base::Quantity;
 
+#define unpack_array(a) &#a[0], &#a[1], &#a[2], &#a[3], &#a[4], &#a[5], &#a[6], &#a[7]
```
Is there a way to express this C++ way?
See also https://github.com/llvm/llvm-project/pull/137577 and #21019. Both would also help here.

hyarion replied:
> Is there a way to express this C++ way?

Short answer: Yes

Long answer:
```cpp
#include <array>
#include <iostream>
#include <tuple>
#include <utility>
#include <type_traits>

// Mock python api to
using PyObject = std::tuple<double, int, int, int, int, int, int, int, int>;
bool PyArg_ParseTuple(PyObject args, const char* fmt, double* f,
                      int* a0, int* a1, int* a2, int* a3,
                      int* a4, int* a5, int* a6, int* a7) {
    if (fmt != std::string("|diiiiiiii")) return false;
    *f = std::get<0>(args);
    *a0 = std::get<1>(args);
    *a1 = std::get<2>(args);
    *a2 = std::get<3>(args);
    *a3 = std::get<4>(args);
    *a4 = std::get<5>(args);
    *a5 = std::get<6>(args);
    *a6 = std::get<7>(args);
    *a7 = std::get<8>(args);
    return true;
}

// Creates a tuple of pointers to array elements using index sequence
// e.g. (&arr[0], &arr[1], ..., &arr[N-1])
template <typename T, std::size_t N, std::size_t... Is>
constexpr auto unpack_ptrs_impl(std::array<T, N>& arr, std::index_sequence<Is...>) {
    return std::tuple(&arr[Is]...);
}

// Generate indices and calls the implementation
template <typename T, std::size_t N>
constexpr auto unpack_ptrs(std::array<T, N>& arr) {
    return unpack_ptrs_impl(arr, std::make_index_sequence<N>{});
}

// Calls the given function with element pointers unpacked from the array
// Useful for C-style APIs that take individual pointers as args
template <typename Func, typename T, std::size_t N>
auto apply_array_ptrs(std::array<T, N>& arr, Func&& func) {
    return std::apply(std::forward<Func>(func), unpack_ptrs(arr));
}

// Example of how to use it
int main() {
    std::array<int, 8> i{};
    double f = 0.0;

    PyObject mockArgs = {123.4, 1, 2, 3, 4, 5, 6, 7, 8};

    // Automatically unpacks array element pointers into PyArg_ParseTuple
    bool success = apply_array_ptrs(i, [&](auto*... ptrs) {
        return PyArg_ParseTuple(mockArgs, "|diiiiiiii", &f, ptrs...);
    });

    if (!success) {
        std::cerr << "Failed to parse arguments.\n";
        return 0;
    }
    std::cout << "Parsed float: " << f << "\n";
    std::cout << "Parsed ints: ";
    for (int val : i) {
        std::cout << val << ' ';
    }
    std::cout << "\n";

    return 0;
}
```

3x380V replied:
Eh... That's quite noisy, thus indeed long :)

hyarion replied:
Yeah, but it's not as bad if you would split up the templates:
```c++
template <typename T, std::size_t N, std::size_t... Is>
constexpr auto unpack_ptrs_impl(std::array<T, N>& arr, std::index_sequence<Is...>) {
    return std::tuple(&arr[Is]...);
}

template <typename T, std::size_t N>
constexpr auto unpack_ptrs(std::array<T, N>& arr) {
    return unpack_ptrs_impl(arr, std::make_index_sequence<N>{});
}

template <typename Func, typename T, std::size_t N>
auto apply_array_ptrs(std::array<T, N>& arr, Func&& func) {
    return std::apply(std::forward<Func>(func), unpack_ptrs(arr));
}
```
from how to use it:
```c++
    bool success = apply_array_ptrs(i, [&](auto*... ptrs) {
        return PyArg_ParseTuple(mockArgs, "|diiiiiiii", &f, ptrs...);
    });
```

I'm not quite sure which should be preferred though...

3x380V replied:
@bofdahof, `#define` is just an experiment to fool auto formatting.

Those values technically are exponents, it is just  _T^α ∙ L^β ∙ M^γ ∙ I^δ ∙ Θ^ε ∙ N^ζ ∙ J^η_, where _α, β, γ, δ, ε, ζ, η_ are the unit `vals` also known as exponents. I have no problem with either.

3x380V replied:
See also https://github.com/FreeCAD/FreeCAD/pull/21147/commits/0be3772b9857b5f23cbb173f9ba7d09ccf417fe0. Of course I'd like to hear objections, etc.

3x380V replied:
Expression is correct ;-)

Consistency, yes... Any consistent proposal?

hyarion replied:
`vals` is a bit too generic for my taste. A bit like naming an int `number`.

The correct name for these values are "dimensional exponents" (2.3.3 p136). I usually go with long declarative names so I would probably name it `dimensionalExponents` but `dimExponents` or even `dimExps` would also work depending on what you prefer.

I don't think it needs to be changed in this PR but feel free to do it if you like

3x380V replied:
For a reference it is pdf page 22 of [SI Brochure](https://www.bipm.org/documents/20126/41483022/SI-Brochure-9-EN.pdf). I have no strong preference here, so I'll wait for @bofdahof's input.

3x380V replied:
Ok, since there is an agreement on `exps`, lets use that. But unless anyone wants step in, development will slow down now because of the regular work...

hyarion replied:
As mentioned earlier, renaming `vals` is a nice bonus, but can be omitted in this PR if you like.

Doing it in this PR will most likely require more work to keep the PR up to date and collision free.

Renaming it in a later PR with just that change will be a small, easy to review PR which can be merged swiftly.

I'm just worried that this PR drags on and will cause you more pain than necessary. But it's your choice.

3x380V replied:
Fair enough.

hyarion replied:
Regarding this change, this can be left for the future to be cleaned up. Maybe bof's original version is simple enough for now?

3x380V replied:
> > added back OverflowError and UnderflowError
> 
> Why is that necessary?

This question deserves a bit longer answer.
The original author considered thrown exception so important, he even write test for it:
```c++
TEST(Unit, TestUnderflow)
{
    // this tests _that_ the expected exception is thrown
    EXPECT_THROW(
        {
            try {
                Base::Unit unit {3};
                unit.pow(-10000);
            }
            catch (const Base::UnderflowError& e) {
                // and this tests that it has the correct message
                EXPECT_STREQ("Unit underflow in pow()", e.what());
                throw;
            }
        },
        Base::UnderflowError);
}
```
This is visible also from Python code and based on the recent experience with amount of regressions reported from Python side of things I'm reluctant changing that behavior even if that would mean slightly more complicated code. There are only few Python tests and they almost certainly cover only small minority of use cases, so it is hard to guess what might break. Hopefully this also explain my concern to merge simple enough version.

3x380V commented on src/Base/QuantityPyImp.cpp:
```diff
@@ -105,27 +107,14 @@ int QuantityPy::PyInit(PyObject* args, PyObject* /*kwd*/)
+    if (PyArg_ParseTuple(args, "|diiiiiiii", &f, unpack_array(i))) {
+        UnitVals vals {};
+        std::transform(i.begin(), i.end(), vals.begin(), [&](auto val) {
+            return static_cast<int8_t>(val);
```
Should we `throw` here for number exceeding 8bit integer range? `0xff01` would silently become 1, perfectly valid exponent.

3x380V commented:
I'm a bit confused by `QuantityPy` methods. Some of them are catching exceptions and returning `-1`, some of them just `throw`. As a few changes happened in this class, PR stays in Draft mode until investigated. What a great potential to unexpectedly break things! All this Units stuff much like reminds the [1172](https://xkcd.com/1172/) misfortune.

3x380V commented:
Resembled original way to handle parsing. Now exception is caught and error returned, so it should behave the same way as before. That said, PR is ready for broader testing, I'm just unsure how to convince people to test before merge. The only thing left I'm aware of is that `#define` thing, which I'm not sure is worse than a template.

3x380V commented:
Yes, you saw that right above your comment. I added this with very same exception type and text, so that become obvious there is space for improvement.  Unit itself limits exponent being in (AFAIR) -7, 8 range and that's because there were only 4 bits per value. I added another check so it does not silently truncate integers while typecasting, so using more permissive constructor and do only one check might be quite good solution.

hyarion commented:
@3x380V Sad to see this closed. Is there anything I can help with to get it across the finish line?

3x380V commented:
This PR most likely touches the same code which is the culprit of #21353 and I do not have enough time to deal with both. I also originally opened this one just to resolve conflicts and now we are too far beyond that point. Anyone is welcome to revisit and reopen it, git history is preserved here and I'm not going to delete feature branch.

3x380V commented:
> Somewhere I saw discussion about range checking unit EXPONENT (!!) values coming from python and see there was additional code added in response. The down side is now range checking done in two places: python interface and unit. Unit expects a 8 bit integer type. Maybe that is too restrictive and its constructor should be more permissive and cast as required (which is also being done in python interface) and use the range check function in Unit as the source of truth.

Added Unit constructor taking those integers. This is a helper one to ease constructing Units from Python, so we follow KISS. I certainly don't hesitate implementing whatever cleaner solution suggested.
As Python is quite prominent user of Unit code and Unit code still behave a bit differently than before, this stays in Draft mode.

> The original was fine without many of the changes that have been grafted in and for me the PR is unrecongnizable and I really don't want to participate in this further.

@bofdahof, I did my best to incorporate your changes into ever moving code base while taking review comments and problems I hit while using it into an account. I'm sorry you find it too disruptive.

3x380V commented:
> Somewhere I saw discussion about range checking unit EXPONENT (!!) values coming from python and see there was additional code added in response. The down side is now range checking done in two places: python interface and unit. Unit expects a 8 bit integer type. Maybe that is too restrictive and its constructor should be more permissive and cast as required (which is also being done in python interface) and use the range check function in Unit as the source of truth.

Added Unit constructor taking those integers. This is a helper one to ease constructing Units from Python, so we follow KISS. I certainly don't hesitate implementing whatever cleaner solution suggested.
As Python is quite prominent user of Unit code and Unit code still behave a bit differently than before, this stays in Draft mode.

> The original was fine without many of the changes that have been grafted in and for me the PR is unrecongnizable and I really don't want to participate in this further.

@bofdahof, I did my best to incorporate your changes into ever moving code base while taking review comments and problems I hit while using it into an account. I'm sorry you find it too disruptive.

3x380V commented:
After playing with a few models extensively using Assembly4 and TechDraw and finding everything still working, I'm a bit more confident we should give this a try. It should still be reviewed carefully though...

hyarion commented:
I've read through the code again and I think it looks good.
The description still hints that there's still work to do in the last commit, is that still the case?

Testing wise I'm not sure how to proceed. Maybe we could ask some developers, who uses a lot of units, to test it (material and FEM comes to mind). But after that our only way of testing it properly is to merge it to get it into a dev-build.
We're probably mid through a release so it is still ok (though ofc not preferred) to accidentally introduce some regressions.

Syres916 commented:
@3x380V it's going to be tomorrow evening before I can give this a thorough test but something I've noticed with current `main` is the scientific notation letter `E` is capitalised and I thought it was supposed to be lowercase, screenshots from a variety of tasks in CfdOF which I have a set of macros to build an identical benchmark model. (Standard schema being used)

![Screenshot from 2025-05-19 23-16-38](https://github.com/user-attachments/assets/7916058e-8da3-4f31-9719-dd588eb922dd)
![Screenshot from 2025-05-19 23-16-53](https://github.com/user-attachments/assets/a01b0c29-f73b-436d-a6d2-adc78c9d698c)
![Screenshot from 2025-05-19 23-17-31](https://github.com/user-attachments/assets/d36b90e1-37fd-4ffa-8b8e-fe6416c3ac6e)

```
OS: EndeavourOS (X-Cinnamon/cinnamon/xcb)
Architecture: x86_64
Version: 1.1.0dev.41789 (Git)
Build date: 2025/05/19 18:24:17
Build type: Release
Branch: main
Hash: 1cfb85a71f4cd113deeccdc175850a37f2ad779d
Python 3.13.3, Qt 6.9.0, Coin 4.0.3, Vtk 9.4.2, boost 1_88, Eigen3 3.4.0, PySide 6.9.0
shiboken 6.9.0, SMESH 7.7.1.0, xerces-c 3.3.0, OCC 7.8.1
Locale: English/United Kingdom (en_GB)
Navigation Style/Orbit Style/Rotation Mode: Touchpad/Trackball/Drag at cursor
Stylesheet/Theme/QtStyle: Behave-dark.qss/Behave-dark/
Logical DPI/Physical DPI/Pixel Ratio: 96/95.9222/1
Installed mods: 
  * FreeCAD-themes 2025.1.7
  * CfdOF 1.30.4
  * sheetmetal 0.7.22
  * fasteners 0.5.38

```

hyarion commented:
> it's going to be tomorrow evening before I can give this a thorough test

@Syres916 I'll add you as a reviewer so we don't forget, thanks for volunteering!

3x380V commented:
> @3x380V it's going to be tomorrow evening before I can give this a thorough test but something I've noticed with current `main` is the scientific notation letter `E` is capitalised and I thought it was supposed to be lowercase, screenshots from a variety of tasks in CfdOF which I have a set of macros to build an identical benchmark model. (Standard schema being used)

This is a bit strange. It is the result of `UnitsSchema::toLocale` and I still see lowercase »e« here using en_GB.UTF-8 locale
![Screenshot from 2025-05-20 11-31-17](https://github.com/user-attachments/assets/717720f4-1fc8-4d43-ae8d-c990b7258e80)

Syres916 commented:
> This is a bit strange. It is the result of `UnitsSchema::toLocale` and I still see lowercase »e« here using en_GB.UTF-8 locale 

I'm so glad I'm using Snapshots, I'll go back to April 24th which I have access to both Qt6.8.3 and Qt6.9.0 and re-run my checks, I really hope this isn't an Arch based distro issue due to using bleeding edge libraries.

Syres916 commented:
@3x380V one question though, assuming the `InputField` validation is working correctly, do you understand why those two fields in your screenshot are still showing red=invalid input?

3x380V commented:
> ...do you understand why those two fields in your screenshot are still showing red=invalid input?

I haven't investigated too much here, but
- starting to edit value (add space, for example) makes that red sign go away
- selecting another fluid from combobox makes that red sign go away

Syres916 commented:
The red Invalid icon is definitely a regression but nothing to do with #20540 , managed to find an AppImage from before the new Units were merged and a correctly entered field even from a macro has always in the past shown as valid on opening the appropriate task panel, can't believe it's not been reported but don't remember seeing an Issue for it.

Syres916 commented:
I'll test the PR in isolation and then repeat with #21465 just incase my testing provides errors that maybe are fixed by the incorrect ordering.

hyarion commented:
> Hmm, that #21465 seems correct. @hyarion what about merging that as well as #21466? I'll then rebase this PR to ease testing.

Done

Syres916 commented:
Regression detected, after renaming attached file back to .FCStd, open file, change document schema to Building US, mark the entire doc for Recompute and trigger Recompute. The Draft dimension should change to `3.16 in` but instead Report View output:

```
19:58:25  pyException: Traceback (most recent call last):
  File "/home/username/freecad-daily-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
  File "/home/username/freecad-daily-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
<class 'ZeroDivisionError'>: float division by zero
19:58:25  pyException: Traceback (most recent call last):
  File "/home/username/freecad-daily-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
  File "/home/username/freecad-daily-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
<class 'ZeroDivisionError'>: float division by zero
19:58:25  pyException: Traceback (most recent call last):
  File "/home/username/freecad-daily-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
  File "/home/username/freecad-daily-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
<class 'ZeroDivisionError'>: float division by zero
19:58:25  pyException: Traceback (most recent call last):
  File "/home/username/freecad-daily-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
  File "/home/username/freecad-daily-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
<class 'ZeroDivisionError'>: float division by zero
19:58:25  pyException: Traceback (most recent call last):
  File "/home/username/freecad-daily-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
  File "/home/username/freecad-daily-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
<class 'ZeroDivisionError'>: float division by zero
19:58:25  pyException: Traceback (most recent call last):
  File "/home/username/freecad-daily-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
  File "/home/username/freecad-daily-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
<class 'ZeroDivisionError'>: float division by zero
```

[DraftDimensionTest.FCStd.zip](https://github.com/user-attachments/files/20353962/DraftDimensionTest.FCStd.zip)

3x380V commented:
@Syres916, cannot reproduce. Recomputes and dimension become 3 1/8"

Syres916 commented:
@3x380V in order for me to understand what's going on, could you please replace the `display_external` function in your local src/Mod/Draft/draftutils/units.py with:

```
def display_external(internal_value,
                     decimals=None, dim='Length', showUnit=True, unit=None):
    """Return a converted value for display, according to the unit schema.

    Parameters
    ----------
    internal_value: float
        A value that will be transformed depending on the other parameters.

    decimals: float, optional
        It defaults ot `None`, in which case, the decimals are 2.

    dim: str, optional
        It defaults to `'Length'`. It can also be `'Angle'`.

    showUnit: bool, optional
        It defaults to `True`.
        If it is `False` it won't show the unit.

    unit: str, optional
        A unit string such as `'mm'`, `'cm'`, `'m'`, `'in'`, `'ft'`,
        in which to express the returned value.
    """
    print("unit: " + str(unit))
    print("decimals: " + str(decimals))
    if dim == 'Length':
        q = App.Units.Quantity(internal_value, App.Units.Length)
        print("q: " + str(q))
        if not unit:
            if decimals is None and showUnit:
                return q.UserString

            conversion = q.getUserPreferred()[1]
            uom = q.getUserPreferred()[2]
        elif unit.lower() == "arch":
            print("It is an arch unit")
            return App.Units.schemaTranslate(q,5)[0].replace("+"," ")
        else:
            try:
                uom = unit
                internal_value = q.getValueAs(unit)
                conversion = 1
            except Exception:
                conversion = q.getUserPreferred()[1]
                uom = q.getUserPreferred()[2]
    elif dim == 'Angle':
        q = App.Units.Quantity(internal_value, App.Units.Angle)
        if decimals is None:
            return q.UserString

        conversion = q.getUserPreferred()[1]
        uom = q.getUserPreferred()[2]
    else:
        conversion = 1.0
        if not decimals:
            decimals = 2
        uom = "??"

    if not showUnit:
        uom = ""
    print("conversion: " + str(float(conversion)))
    print("uom: " + str(uom))
    decimals = abs(decimals)  # prevent negative values
    fmt = "{0:." + str(decimals) + "f} " + uom
    print("fmt: " + str(fmt))
    print("internal_value: " + str(float(internal_value)))
    display = fmt.format(float(internal_value) / float(conversion))
    display = display.replace(".", QtCore.QLocale().decimalPoint())

    return display
```

and then redo the test on my DraftDimensionTest file and please post the Report View output. I'm really struggling to believe my Locale can be causing the `conversion` to be 0.0.

3x380V commented:
On the file load with Standard schema selected it displays:
<details><summary>Standard schema</summary>

```
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: None
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 1.0
uom: mm
fmt: {0:.2f} mm
internal_value: 80.32749999999999
```

</details>

![Screenshot from 2025-05-21 08-08-50](https://github.com/user-attachments/assets/d73066f6-8889-46f3-bf80-393f9a03a7db)

Changing schema to US Customary
<details><summary>US Customary</summary>

```
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
```

</details>

![Screenshot from 2025-05-21 08-09-09](https://github.com/user-attachments/assets/b73ae148-9f7e-482b-bced-5c37bbac8e59)

Changing schema to Building US:
<details><summary>Building US</summary>

```
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
unit: 
decimals: 2
q: 80.32749999999999 mm
conversion: 25.4
uom: "
fmt: {0:.2f} "
internal_value: 80.32749999999999
```

</details>

![Screenshot from 2025-05-21 08-15-53](https://github.com/user-attachments/assets/b06aca43-1c79-4cfa-9626-bec220aa2cff)

Also tried to change locale to  en_GB.ISO-8859-15, which worked the same apart from:
<details><summary>Locale error</summary>

```
xkbcommon: ERROR: /usr/share/X11/locale/iso8859-15/Compose:47:29: string literal is not a valid UTF-8 string
[...]
xkbcommon: ERROR: /usr/share/X11/locale/iso8859-15/Compose:48:29: string literal is not a valid UTF-8 string
xkbcommon: ERROR: /usr/share/X11/locale/iso8859-15/Compose:48:29: too many errors
xkbcommon: ERROR: /usr/share/X11/locale/iso8859-15/Compose:48:29: failed to parse file
```

</details>

3x380V commented:
Fixed Units convertor. It is now broken the same way as in `main` branch. Probably another regression, will investigate later.

Syres916 commented:
@FEA-eng it would be really helpful if you can give this PR a test from a FEM perspective as there's a significant portion of Python in the workbench. A Qt6 based distro would be the best but if you've only got Qt5 then that will be better than no feedback.

If you get the red invalid input icon (such as https://github.com/FreeCAD/FreeCAD/pull/21147#issuecomment-2893656914) in any task panels just ignore them, that's a non-Units related issue, what's important is that what you see in the fields being correct, basically in CfdOF all that I can check is showing correctly but all are showing invalid on opening which I'll deal with separately.

Syres916 commented:
@3x380V I spun up a new distro VM Linux Mint 22.1, went through the whole palaver to get it compiling and then compiled you PR, repeated the same test with the Darft>Dimension and exactly the same error:

```
10:45:08  pyException: Traceback (most recent call last):
  File "/home/john/freecad-units2-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/john/freecad-units2-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
<class 'ZeroDivisionError'>: float division by zero
10:45:08  pyException: Traceback (most recent call last):
  File "/home/john/freecad-units2-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/john/freecad-units2-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
<class 'ZeroDivisionError'>: float division by zero
10:45:08  pyException: Traceback (most recent call last):
  File "/home/john/freecad-units2-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/john/freecad-units2-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
<class 'ZeroDivisionError'>: float division by zero
10:45:08  pyException: Traceback (most recent call last):
  File "/home/john/freecad-units2-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/john/freecad-units2-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
<class 'ZeroDivisionError'>: float division by zero
10:45:08  pyException: Traceback (most recent call last):
  File "/home/john/freecad-units2-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/john/freecad-units2-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
<class 'ZeroDivisionError'>: float division by zero
10:45:08  pyException: Traceback (most recent call last):
  File "/home/john/freecad-units2-build/Mod/Draft/draftviewproviders/view_dimension.py", line 611, in updateData
    self.string = units.display_external(length,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/john/freecad-units2-build/Mod/Draft/draftutils/units.py", line 131, in display_external
    display = fmt.format(float(internal_value) / float(conversion))
                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
<class 'ZeroDivisionError'>: float division by zero
```



```
OS: Linux Mint 22.1 (X-Cinnamon/cinnamon/xcb)
Architecture: x86_64
Version: 1.1.0dev.41806 (Git)
Build date: 2025/05/21 10:28:36
Build type: Release
Branch: More_Units_Testing
Hash: ae6fadd1f8031b05554224b2ad80a7199a50f3a2
Python 3.12.3, Qt 5.15.13, Coin 4.0.2, Vtk 9.1.0, boost 1_83, Eigen3 3.4.0, PySide 5.15.13
shiboken 5.15.13, SMESH 7.7.1.0, xerces-c 3.2.4, IfcOpenShell 0.8.2, OCC 7.6.3
Locale: English/United Kingdom (en_GB)
Navigation Style/Orbit Style/Rotation Mode: Touchpad/Trackball/Drag at cursor
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/Qt default
Logical DPI/Physical DPI/Pixel Ratio: 96/95.9245/1
Installed mods: 
  * CfdOF 1.30.4

```

3x380V commented:
@Syres916, what is ae6fadd1f8031b05554224b2ad80a7199a50f3a2? Looks familiar, but I'm unable to tell whenever there are any other modifications.
Anyway, the only way `conversion` may be zero is that `QuantityPy::getUserPreferred` returns that. So `QuantityPy::getUserPreferred` → `Quantity::getUserString` → `UnitsApi::schemaTranslate` → `UnitsSchema::translate` where `translate` need to be debugged. Can you do this?

Syres916 commented:
> @Syres916, what is [ae6fadd](https://github.com/FreeCAD/FreeCAD/commit/ae6fadd1f8031b05554224b2ad80a7199a50f3a2)?

It's simply your branch PR'd into a clean `main` , see https://github.com/Syres916/FreeCAD/tree/More_Units_Testing so  I have a non-moving goal post to work with.


 > Anyway, the only way `conversion` may be zero is that `QuantityPy::getUserPreferred` returns that. So `QuantityPy::getUserPreferred` → `Quantity::getUserString` → `UnitsApi::schemaTranslate` → `UnitsSchema::translate` where `translate` need to be debugged. Can you do this?

I'll have a try but I'm not going to get any serious time until late Saturday or Sunday and I've just agreed to write a Unit Test for BIM IFC. It might be far easier for you to spin up a Linux Mint 22.1 VM and install it with English keyboard, English language and London Time Zone with Cinnamon Desktop so you know it's built exactly the same way as mine.

yorikvanhavre commented:
Too long to understand by my feeble mind :) Waiting for you guys to resolve this one :)

hyarion commented:
@3x380V Did you manage to reproduce the issue, or do you need help doing that to move this forward?

3x380V commented:
I just rebased and build this PR. Units converter is working again, but I didn't check which change fixed that. Unfortunately I haven't found enough time to reproduce @Syres916's issue.

3x380V commented:
@Syres916, is there any chance you'll be able to re-test any time soon?

3x380V commented:
I took an opportunity and ported two @wwmayer's `Quantity` testing patches meanwhile. I'm aware it could be further simplified using labdas, but decided not to do so. This PR is never ending story enough already.

Syres916 commented:
@3x380V I'm catching up, hopefully I'll have some feedback in the next couple of days.

3x380V commented:
@hyarion, although I was not able to reproduce @Syres916's crash, `factor` (that's `conversion` in Python code) was not explicitly initialized in some cases, which may lead to the observed crash.
There is nothing specific I'd like to test, most of bug reports come from Python code, which I was not even aware exist :)
